samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -1,276 +1,215 @@
+        """
+        Extracts an api id from an HTTP or REST api arn.
+
+        NOTE: https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html
+
+        REST API Arn: `arn:partition:apigateway:region::/restapis/api-id`
+
+        HTTP API Arn: `arn:partition:apigateway:region::/apis/api-id`
+
+        Parameters
+        ----------
+        api_arn : str
+            The arn of the HTTP or REST api from which the api id is extracted.
+
+        Returns
+        -------
+        str
+            The api id from the api arn.
+        """
+
+        return api_arn[api_arn.rindex("/") + 1 :]
+
+    def _create_iam_statement_string(self, resource_arn: str) -> Union[dict, None]:
+        """
+        Returns an IAM Statement in the form of a YAML string corresponding to the supplied resource ARN.
+
+        The generated Actions are commented out to establish a barrier of consent, preventing customers from accidentally granting permissions that they did not mean to.
+
+        The Action list is empty if there are no IAM Action permissions to generate for the given resource.
+
+        Returns `None` if the `resource_arn` corresponds to an IAM or STS resource.
+
+        Parameters
+        ----------
+        resource_arn : str
+            The arn of the resource for which the IAM statement is generated.
+
+        Returns
+        -------
+        str:
+            An IAM Statement in the form of a YAML string.
+
+            `None` if the `resource_arn` corresponds to an IAM or STS resource.
+
+        Raises
+        ------
+        jinja2.exceptions.TemplateError:
+            If the IAM Statement template render fails.
+        """
+
+        # We keep partition and region general to avoid the burden of maintaining new regions & partitions.
+        partition_regex = r"[\w-]+"
+        region_regex = r"[\w-]+"
+        account_regex = r"\d{12}"
+
+        # We want to explcitly avoid generating permissions for IAM or STS resources
+        IAM_PREFIX_REGEX = rf"^arn:{partition_regex}:(iam|sts)"
+
+        if re.search(IAM_PREFIX_REGEX, resource_arn):
+            LOG.debug("ARN `%s` is for an IAM or STS resource, will not generate permissions.", resource_arn)
+            return None
+
+        LAMBDA_FUNCTION_REGEX = rf"^arn:{partition_regex}:lambda:{region_regex}:{account_regex}:function:[\w-]+(:\d+)?$"
+        APIGW_API_REGEX = rf"^arn:{partition_regex}:apigateway:{region_regex}::\/apis\/\w+$"
+        APIGW_RESTAPI_REGEX = rf"^arn:{partition_regex}:apigateway:{region_regex}::\/restapis\/\w+$"
+        SQS_QUEUE_REGEX = rf"^arn:{partition_regex}:sqs:{region_regex}:{account_regex}:[\w-]+$"
+        S3_BUCKET_REGEX = rf"^arn:{partition_regex}:s3:::[\w-]+$"
+        DYNAMODB_TABLE_REGEX = rf"^arn:{partition_regex}:dynamodb:{region_regex}:{account_regex}:table\/[\w-]+$"
+        STEPFUNCTION_REGEX = rf"^arn:{partition_regex}:states:{region_regex}:{account_regex}:stateMachine:[\w-]+$"
```

Comment:
This is slightly concerning. Why do we need to maintain regex's here? This is also brittle, as (if) new partitions are build, this requires us to have to maintain that map.

Reply:
Yes, I agree it is definitely not ideal.

I can send the design document for this component offline, but the idea is to make writing IAM statements a bit easier by having a group of commonly used resources for which the template will include a set of commonly used actions. For example, `Lambda` would have `InvokeFunction`.

Which resources to generate IAM statements for is determined via tags (to support CFN & Terraform), so by querying tagging API, we get a list of resource ARNs. Determining which ARN belongs to which resource requires some logic when some arns are subsets of others, e.g. [APIGW ARNs](https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html). It was decided that regex was most pragmatic to link resource ARNs to their type.

As for the partitions, their expressions are general, and should not require changing in the future. I left them in the `partition_regex` variable for clarity.

There's also no real "obligation" to maintain this list of supported resources, it's really just to include the (estimated) most commonly used resources for integration testing.

Reply:
I will ping you off this. I don't fully understand what is going on here. 

> There's also no real "obligation" to maintain this list of supported resources,

I disagree. Once you release something, you (or the team) signs up to maintain it. So if arns change (for whatever reason), we have to change it. So it is something we have to maintain.

Reply:
That was poorly worded, I meant that we don't necessarily need to _add_ resources, the tool will not break if we don't. But yes, we definitely are signing up to maintain any changes to the arns, and even to add new resources to the pool. I didn't intend to say it did not require maintenance, it does.

Reply:
Update: Obviously regex is not ideal. An ideal solution would be a proper ARN parser available in boto3 (Go and Java SDKs have ARN libraries, but not boto), or a tag API that returns resource types as well as arns for resources matching a given tag.

Neither of those will be available any time soon, so I'm going to abstract away as much of the regex to something that can easily be changed later. Once more important aspects of the project are finished and merged, I can go back and replace the mechanism responsible for matching ARNs to IAM statements with a simple ARN parser.